pixy/kafka.go

```diff
@@ -0,0 +1,272 @@
+
+// Produce submits a message to the specified topic of the Kafka cluster using
+// `key` to navigate the message to a particular shard.
+//
+// An attempt to call `Produce` after the client stop has been triggered will
+// result in panic.
+func (kc *KafkaClient) Produce(topic string, key, message []byte) {
+	// Preserve `nil` values while converting from `[]byte` to `sarama.Encoder`.
+	var keyEncoderOrNil sarama.Encoder
+	if key != nil {
+		keyEncoderOrNil = sarama.ByteEncoder(key)
+	}
+
+	prodMsg := &sarama.ProducerMessage{
+		Topic: topic,
+		Key:   keyEncoderOrNil,
+		Value: sarama.ByteEncoder(message),
+	}
+	kc.dispatcherCh <- prodMsg
+}
+
+// merger receives both acknowledged messages and producer errors from the
+// respective `sarama.Producer` channels, constructs `ProducerResult`s out of
+// them and sends the constructed `producerResult` instances to `resultCh` to
+// be further inspected by the `dispatcher` goroutine.
+//
+// It keeps running until both `sarama.Producer` output channels are closed,
+// Then it closes the `KafkaClient.resultCh` to notify the `dispatcher`
+// goroutine that all pending messages are processed and exits.
+func (kc *KafkaClient) merger() {
+	prodSuccessesCh := kc.wrappedProducer.Successes()
+	prodErrorsCh := kc.wrappedProducer.Errors()
+mergeLoop:
+	for channelsOpened := 2; channelsOpened > 0; {
+		select {
+		case ackedMsg, channelOpened := <-prodSuccessesCh:
+			if !channelOpened {
+				channelsOpened -= 1
+				prodSuccessesCh = nil
+				continue mergeLoop
+			}
+			kc.resultCh <- &producerResult{msg: ackedMsg}
+		case prodErr, channelOpened := <-prodErrorsCh:
+			if !channelOpened {
+				channelsOpened -= 1
+				prodErrorsCh = nil
+				continue mergeLoop
+			}
+			kc.resultCh <- &producerResult{msg: prodErr.Msg, err: prodErr.Err}
+		}
+	}
+	// Close the result channel to notify the `dispatcher` goroutine that all
+	// pending messages have been processed.
+	close(kc.resultCh)
+}
+
+// dispatcher is the main `KafkaClient` goroutine. It receives messages from
+// `dispatchedCh` where they are send to by `Produce` method and submits them
+// to `sarama.Producer` for actual delivery. The dispatcher main goal is to
+// ensure that buffered messages are not lost during graceful shutdown. It
+// achieves that by allowing a time period after it stops receiving messages
+// via `dispatchCh` and when the `sarama.Producer` stop is triggered.
+func (kc *KafkaClient) dispatcher() {
+	dispatcherCh := kc.dispatcherCh
+	var prodInputCh chan<- *sarama.ProducerMessage
+	pendingMsgCount := 0
+	// The normal operation loop is implemented as two-stroke machine. On the
+	// first stroke a message is received from `dispatchCh`, and on the second
+	// it is sent to `prodInputCh`. Note that producer results can be received
+	// at any time.
+	prodMsg := (*sarama.ProducerMessage)(nil)
+	channelOpened := true
+	for {
+		select {
+		case prodMsg, channelOpened = <-dispatcherCh:
+			if !channelOpened {
+				goto gracefulShutdown
```

Comment:
You could accomplish the same thing with a labeled break statement. Same with the one below. I don't feel strongly about it, though.